File: Axiom/Internal/Parser.cs
using Axiom.Internal.Ast;
using System;
using System.Collections.Generic;
using System.Text;

namespace Axiom.Internal
{
    internal class Parser : IParser
    {
        private static readonly Token[] AssignOperators = {
            Token.Assign,
            Token.AddAssign,
            Token.Subtract,
            Token.MultiplyAssign,
            Token.DivideAssign,
            Token.ModuloAssign,
            Token.BitwiseAndAssign,
            Token.BitwiseOrAssign,
            Token.BitwiseXorAssign,
            Token.LeftShiftAssign,
            Token.RightShiftAssign,
            Token.BitwiseUnsignedRightShiftAssign
        };

        private static readonly Token[] RelationalOperators = {
            Token.LogicalLess,
            Token.LogicalLessOrEqual,
            Token.LogicalGreater,
            Token.LogicalGreaterOrEqual,
            Token.LogicalEqual,
            Token.LogicalNotEqual
        };

        private ILexer _lexer;
        private Symbol _currentSymbol, _nextSymbol;

        public Parser(ILexer lexer)
        {
            _lexer = lexer;
        }

        public AstNode Parse()
        {
            _nextSymbol = _lexer.Next();

            var statements = ParseStatements();

            return new FunctionExpression(new List<Identifier>(), new BlockStatement(statements));
        }

        #region Statements
        private IList<Statement> ParseStatements()
        {
            var statements = new List<Statement>();

            while (!Match(Token.Eof)) {
                var statement = ParseStatement();

                if (statement != null) {
                    statements.Add(statement);
                }
            }

            return statements;
        }

        private Statement ParseStatement()
        {
            switch (Peek()) {
            case Token.OpenBrace:
                return ParseBlockStatement();
            case Token.OpenParenthesis:
                return ParseExpressionStatement();
            case Token.KeywordIf:
                return ParseIfStatement();
            case Token.KeywordWhile:
                return ParseWhileStatement();
            case Token.KeywordFor:
                return ParseForStatement();
            case Token.SemiColon:
                Accept(Token.SemiColon);

                return null;
            }

            return ParseExpressionStatement();
        }

        private Statement ParseIfStatement()
        {
            Check(Token.KeywordIf);

            Check(Token.OpenParenthesis);
            var condition = ParseExpression();
            Check(Token.CloseParenthesis);
            var body = ParseStatement();

            if (Accept(Token.KeywordElse)) {
                var elseBody = ParseStatement();

                return new IfStatement(condition, body, elseBody);
            }

            return new IfStatement(condition, body);
        }

        private Statement ParseWhileStatement()
        {
            Check(Token.KeywordWhile);

            Check(Token.OpenParenthesis);
            var condition = ParseExpression();
            Check(Token.CloseParenthesis);

            var body = ParseStatement();

            return new WhileStatement(condition, body);
        }

        private Statement ParseForStatement()
        {
            return new ForStatement(null, null, null, null);
        }

        private Statement ParseBlockStatement()
        {
            Check(Token.OpenBrace);

            var statements = ParseStatementList();

            Check(Token.CloseBrace);

            return new BlockStatement(statements);
        }

        private IList<Statement> ParseStatementList()
        {
            var list = new List<Statement>();

            while (true) {
                if (Match(Token.CloseBrace)) {
                    break;
                }

                var statement = ParseStatement();

                if (statement == null) {
                    break;
                }

                list.Add(statement);
            }

            return list;
        }

        private Statement ParseExpressionStatement()
        {
            var stmt = new ExpressionStatement(ParseExpression());

            Accept(Token.SemiColon);

            return stmt;
        }
        #endregion

        #region Expressions
        private Expression ParseExpression()
        {
            var expr = ParseAssignmentExpression();

            if (Match(Token.Comma)) {
                var expressions = new List<Expression>();

                expressions.Add(expr);

                while (Accept(Token.Comma)) {
                    expressions.Add(ParseExpression());
                }

                return new SequenceExpression(expressions);
            }

            return expr;
        }

        private Expression ParseAssignmentExpression()
        {
            var expr = ParseConditionalExpression();

            if (Accept(AssignOperators)) {
                var left = expr;
                var oper = _currentSymbol.Token;
                var right = ParseAssignmentExpression();

                return new AssignmentExpression(oper, left, right);
            }

            return expr;
        }

        private Expression ParseConditionalExpression()
        {
            var expr = ParseOrExpression();

            if (Accept(Token.QuestionMark)) {
                var consequent = ParseAssignmentExpression();
                Check(Token.Colon);
                var alternate = ParseAssignmentExpression();

                return new ConditionalExpression(expr, consequent, alternate);
            }

            return expr;
        }

        private Expression ParseOrExpression()
        {
            var expr = ParseAndExpression();

            if (Accept(Token.LogicalOr)) {
                return new BinaryExpression(Token.LogicalOr, expr, ParseOrExpression());
            }

            return expr;
        }

        private Expression ParseAndExpression()
        {
            var expr = ParseBitwiseOrExpression();

            if (Accept(Token.LogicalAnd)) {
                return new BinaryExpression(Token.LogicalAnd, expr, ParseAndExpression());
            }

            return expr;
        }

        private Expression ParseBitwiseOrExpression()
        {
            var expr = ParseBitwiseXorExpression();

            if (Accept(Token.BitwiseOr)) {
                return new BinaryExpression(Token.BitwiseOr, expr, ParseBitwiseOrExpression());
            }

            return expr;
        }

        private Expression ParseBitwiseXorExpression()
        {
            var expr = ParseBitwiseAndExpression();

            if (Accept(Token.BitwiseXor)) {
                return new BinaryExpression(Token.BitwiseXor, expr, ParseBitwiseXorExpression());
            }

            return expr;
        }

        private Expression ParseBitwiseAndExpression()
        {
            var expr = ParseEqualityExpression();

            if (Accept(Token.BitwiseAnd)) {
                return new BinaryExpression(Token.BitwiseAnd, expr, ParseBitwiseAndExpression());
            }

            return expr;
        }

        private Expression ParseEqualityExpression()
        {
            var expr = ParseRelationalExpression();

            if (Accept(Token.LogicalEqual, Token.LogicalNotEqual)) {
                return new BinaryExpression(_currentSymbol.Token, expr, ParseEqualityExpression());
            }

            return expr;
        }

        private Expression ParseRelationalExpression()
        {
            var expr = ParseShiftExpression();

            if (Accept(RelationalOperators)) {
                return new BinaryExpression(_currentSymbol.Token, expr, ParseRelationalExpression());
            }

            return expr;
        }

        private Expression ParseShiftExpression()
        {
            var expr = ParseAdditiveExpression();

            if (Accept(Token.BitwiseLeftShift, Token.BitwiseRightShift, Token.BitwiseUnsignedRightShift)) {
                return new BinaryExpression(_currentSymbol.Token, expr, ParseShiftExpression());
            }

            return expr;
        }

        private Expression ParseAdditiveExpression()
        {
            var expr = ParseMultiplicativeExpression();

            if (Accept(Token.Add, Token.Subtract)) {
                return new BinaryExpression(_currentSymbol.Token, expr, ParseAdditiveExpression());
            }

            return expr;
        }

        private Expression ParseMultiplicativeExpression()
        {
            var expr = ParseUnaryExpression();

            if (Accept(Token.Multiply, Token.Divide, Token.Modulo)) {
                return new BinaryExpression(_currentSymbol.Token, expr, ParseMultiplicativeExpression());
            }

            return expr;
        }

        private Expression ParseUnaryExpression()
        {
            if (Accept(Token.Increment, Token.Decrement)) {
                var t = _currentSymbol.Token;
                var expr = ParseUnaryExpression();

                if (!IsLeftHandSide(expr)) {
                    Error.Report(expr.Position, "Illegal lhs value");
                }

                return new UnaryExpression(t, expr);
            }

            if (Accept(Token.Subtract, Token.BitwiseNot, Token.LogicalNot, Token.KeywordDelete)) {
                var t = _currentSymbol.Token;
                var expr = ParseUnaryExpression();

                return new UnaryExpression(t, expr);
            }

            return ParsePostfixExpression();
        }

        private Expression ParsePostfixExpression()
        {
            var expr = ParseLeftHandSideExpressionAllowCall();

            if ((Match(Token.Increment, Token.Decrement)) && Peek() != Token.SemiColon) {
                if (!IsLeftHandSide(expr)) {
                    Error.Report(expr.Position, "Illegal lhs value");
                }

                Accept(Token.Increment, Token.Decrement);

                expr = new UnaryExpression(_currentSymbol.Token, expr, true);
            }

            return expr;
        }

        private Expression ParseLeftHandSideExpression()
        {
            var expr = ParsePrimaryExpression();

            while (Match(Token.Dot, Token.OpenBracket)) {
                if (Match(Token.OpenBracket)) {
                    var member = ParseComputedMember();

                    expr = new MemberExpression(MemberExpression.AccessType.ListAccess, expr, member);
                } else {
                    var property = ParseNonComputedMember();

                    expr = new MemberExpression(MemberExpression.AccessType.Property, expr, property);
                }
            }

            return expr;
        }

        private Expression ParseLeftHandSideExpressionAllowCall()
        {
            var expr = ParsePrimaryExpression();

            while (Match(Token.Dot, Token.OpenBracket, Token.OpenParenthesis)) {
                if (Match(Token.OpenParenthesis)) {
                    var args = ParseArguments();

                    expr = new CallExpression(expr, args);
                } else if (Match(Token.OpenBracket)) {
                    var member = ParseComputedMember();

                    expr = new MemberExpression(MemberExpression.AccessType.ListAccess, expr, member);
                } else {
                    var property = ParseNonComputedMember();

                    expr = new MemberExpression(MemberExpression.AccessType.Property, expr, property);
                }
            }

            return expr;
        }

        private Expression ParseComputedMember()
        {
            Check(Token.OpenBracket);
            var expr = ParseExpression();
            Check(Token.CloseBracket);

            return expr;
        }

        private Identifier ParseNonComputedMember()
        {
            Check(Token.Dot);

            return ParseNonComputedProperty();
        }

        private Identifier ParseNonComputedProperty()
        {
            Check(Token.Identifier);

            return new Identifier(_currentSymbol.Lexeme);
        }

        private Expression ParseMemberExpression()
        {
            var expr = ParsePrimaryExpression();

            while (true) {
                if (Accept(Token.OpenBracket)) {
                    var left = expr;
                    var right = ParseExpression();

                    Check(Token.CloseBracket);

                    expr = new MemberExpression(MemberExpression.AccessType.ListAccess, left, right);
                } else if (Accept(Token.Dot)) {
                    var left = expr;
                    var right = ParseMemberExpression();

                    expr = new MemberExpression(MemberExpression.AccessType.Property, left, right);
                } else if (Accept(Token.OpenParenthesis)) {
                    var left = expr;
                    var arguments = ParseArguments();

                    Check(Token.CloseParenthesis);

                    if (!(left is Identifier)) {
                        Error.Report(left.Position, "Callable variable expected");
                    }

                    expr = new CallExpression((Identifier)left, arguments);
                } else {
                    break;
                }
            }

            return expr;
        }

        private IList<Expression> ParseArguments()
        {
            var arguments = new List<Expression>();

            while (true) {
                if (Match(Token.CloseParenthesis)) {
                    break;
                }

                arguments.Add(ParseAssignmentExpression());

                if (!Accept(Token.Comma)) {
                    break;
                }
            }

            return arguments;
        }

        private Expression ParsePrimaryExpression()
        {
            if (Accept(Token.OpenParenthesis)) {
                var expr = ParseExpression();
                Check(Token.CloseParenthesis);

                return expr;
            }

            if (Accept(Token.Identifier)) {
                return new Identifier(_currentSymbol.Lexeme);
            }

            if (Accept(Token.KeywordThis)) {
                return new ReferenceExpression(ReferenceExpression.ReferenceType.This);
            }

            if (Accept(Token.KeywordBase)) {
                return new ReferenceExpression(ReferenceExpression.ReferenceType.Base);
            }

            if (Accept(Token.IntegerLiteral)) {
                return new Literal(long.Parse(_currentSymbol.Lexeme), Literal.LiteralType.Integer);
            }

            if (Accept(Token.DecimalLiteral)) {
                return new Literal(double.Parse(_currentSymbol.Lexeme), Literal.LiteralType.Decimal);
            }

            if (Accept(Token.StringLiteral)) {
                return new Literal(_currentSymbol.Lexeme, Literal.LiteralType.String);
            }

            if (Accept(Token.LogicalLiteral)) {
                return new Literal(_currentSymbol.Lexeme == "true", Literal.LiteralType.Logical);
            }

            if (Accept(Token.KeywordNil)) {
                return new Literal(null, Literal.LiteralType.Nil);
            }

            if (Match(Token.Dollar)) {
                return ParseFunctionExpression();
            }

            if (Match(Token.Hash)) {
                return ParseClassExpression();
            }

            if (Match(Token.OpenBrace)) {
                return ParseObjectInitializer();
            }

            Error.Report();

            return null;
        }

        private Expression ParseFunctionExpression()
        {
            Check(Token.Dollar, Token.OpenParenthesis);

            var parameters = ParseParameters();

            Check(Token.CloseParenthesis);

            var body = ParseStatement();

            return new FunctionExpression(parameters, body);
        }

        private Expression ParseClassExpression()
        {
            return null;
        }

        private Expression ParseObjectInitializer()
        {
            return null;
        }

        private IList<Identifier> ParseParameters()
        {
            var parameters = new List<Identifier>();

            while (Accept(Token.Identifier)) {
                parameters.Add(new Identifier(_currentSymbol.Lexeme));

                if (!Accept(Token.Comma)) {
                    break;
                }
            }

            return parameters;
        }
        #endregion

        #region Helpers
        private Token Peek()
        {
            return _nextSymbol.Token;
        }

        private bool Match(Token token)
        {
            return Peek() == token;
        }

        private bool Match(params Token[] tokens)
        {
            foreach (var token in tokens) {
                if (Peek() == token) {
                    return true;
                }
            }

            return false;
        }

        private void Check(Token token)
        {
            if (Match(token)) {
                _currentSymbol = _nextSymbol;
                _nextSymbol = _lexer.Next();
            } else {
                Error.Report(_currentSymbol.Position, "Expected \"{0}\", got \"{1}\"", token, _nextSymbol.Token);
            }
        }

        private void Check(params Token[] tokens)
        {
            foreach (var token in tokens) {
                Check(token);
            }
        }

        private bool Accept(Token token)
        {
            if (Match(token)) {
                _currentSymbol = _nextSymbol;
                _nextSymbol = _lexer.Next();

                return true;
            }

            return false;
        }

        private bool Accept(params Token[] tokens)
        {
            foreach (var token in tokens) {
                if (Accept(token)) {
                    return true;
                }
            }

            return false;
        }

        private bool IsLeftHandSide(Expression expr)
        {
            return expr is Identifier || expr is MemberExpression;
        }
        #endregion

        public void Dispose()
        {
            _lexer.Dispose();
        }
    }
}

File: src/main/java/com/moybl/axiom/Parser.java
package com.moybl.axiom;

import com.moybl.axiom.ast.AssignmentExpression;
import com.moybl.axiom.ast.BinaryExpression;
import com.moybl.axiom.ast.BlockStatement;
import com.moybl.axiom.ast.CallExpression;
import com.moybl.axiom.ast.ConditionalExpression;
import com.moybl.axiom.ast.Expression;
import com.moybl.axiom.ast.ExpressionStatement;
import com.moybl.axiom.ast.FunctionExpression;
import com.moybl.axiom.ast.Identifier;
import com.moybl.axiom.ast.IfStatement;
import com.moybl.axiom.ast.Literal;
import com.moybl.axiom.ast.MemberExpression;
import com.moybl.axiom.ast.Node;
import com.moybl.axiom.ast.ReferenceExpression;
import com.moybl.axiom.ast.SequenceExpression;
import com.moybl.axiom.ast.Statement;
import com.moybl.axiom.ast.UnaryExpression;
import com.moybl.axiom.ast.WhileStatement;

import java.util.ArrayList;
import java.util.List;

public class Parser {

	private static final Token[] ASSIGNMENT_OPERATORS = new Token[]{
			Token.ASSIGN,
			Token.ASSIGN_ADD,
			Token.ASSIGN_SUBTRACT,
			Token.ASSIGN_MULTIPLY,
			Token.ASSIGN_DIVIDE,
			Token.ASSIGN_MODULO,
			Token.ASSIGN_BITWISE_AND,
			Token.ASSIGN_BITWISE_OR,
			Token.ASSIGN_BITWISE_XOR,
			Token.ASSIGN_BITWISE_LEFT_SHIFT,
			Token.ASSIGN_BITWISE_RIGHT_SHIFT,
			Token.ASSIGN_BITWISE_UNSIGNED_RIGHT_SHIFT
	};
	private static final Token[] RELATIONAL_OPERATORS = new Token[]{
			Token.LOGICAL_LESS,
			Token.LOGICAL_LESS_OR_EQUAL,
			Token.LOGICAL_GREATER,
			Token.LOGICAL_GREATER_OR_EQUAL,
			Token.LOGICAL_EQUAL,
			Token.LOGICAL_NOT_EQUAL,
	};

	private Lexer lexer;
	private Symbol current;
	private Symbol next;

	public Parser(Lexer lexer) {
		this.lexer = lexer;
	}

	public Node parse() {
		next = lexer.next();

		return new FunctionExpression(
				new ArrayList<Identifier>(),
				new BlockStatement(parseStatements()));
	}

	private List<Statement> parseStatements() {
		List<Statement> statements = new ArrayList<Statement>();

		while (!match(Token.EOF)) {
			Statement statement = parseStatement();

			if (statement != null) {
				statements.add(statement);
			}
		}

		return statements;
	}

	private Statement parseStatement() {
		switch (next.getToken()) {
			case OPEN_BRACE:
				return parseBlockStatement();
			case OPEN_PARENTHESIS:
				return parseExpressionStatement();
			case KEYWORD_IF:
				return parseIfStatement();
			case KEYWORD_WHILE:
				return parseWhileStatement();
			case KEYWORD_FOR:
				return parseForStatement();
			case SEMI_COLON:
				accept(Token.SEMI_COLON);
				return null;
		}

		return parseExpressionStatement();
	}

	private Statement parseWhileStatement() {
		check(Token.KEYWORD_WHILE);
		Expression condition = parseExpression();
		Statement body = parseStatement();

		return new WhileStatement(condition, body);
	}

	private Statement parseForStatement() {
		return null;
	}

	private Statement parseIfStatement() {
		check(Token.KEYWORD_IF);
		Expression condition = parseExpression();
		Statement body = parseStatement();

		if (accept(Token.KEYWORD_ELSE)) {
			Statement elseBody = parseStatement();

			return new IfStatement(condition, body, elseBody);
		}

		return new IfStatement(condition, body);
	}

	private Statement parseBlockStatement() {
		check(Token.OPEN_BRACE);
		List<Statement> statements = parseStatementList();
		check(Token.CLOSE_BRACE);

		return new BlockStatement(statements);
	}

	private List<Statement> parseStatementList() {
		List<Statement> list = new ArrayList<Statement>();

		while (true) {
			if (match(Token.CLOSE_BRACE)) {
				break;
			}

			Statement statement = parseStatement();

			if (statement == null) {
				break;
			}

			list.add(statement);
		}

		return list;
	}

	private Statement parseExpressionStatement() {
		Statement statement = new ExpressionStatement(parseExpression());
		accept(Token.SEMI_COLON);

		return statement;
	}

	// Expressions
	private Expression parseExpression() {
		Expression expression = parseAssignmentExpression();

		if (match(Token.COMMA)) {
			List<Expression> expressions = new ArrayList<Expression>();
			expressions.add(expression);

			while (accept(Token.COMMA)) {
				expressions.add(parseExpression());
			}

			return new SequenceExpression(expressions);
		}

		return expression;
	}

	private Expression parseAssignmentExpression() {
		Expression expression = parseConditionalExpression();

		if (accept(ASSIGNMENT_OPERATORS)) {
			Token operator = current.getToken();
			Expression destination = parseAssignmentExpression();

			return new AssignmentExpression(operator, expression, destination);
		}

		return expression;
	}

	private Expression parseConditionalExpression() {
		Expression expression = parseOrExpression();

		if (accept(Token.QUESTION_MARK)) {
			Expression consequent = parseAssignmentExpression();
			check(Token.COLON);
			Expression alternate = parseAssignmentExpression();

			return new ConditionalExpression(expression, consequent, alternate);
		}

		return expression;
	}

	private Expression parseOrExpression() {
		Expression expression = parseAndExpression();

		if (accept(Token.LOGICAL_OR)) {
			return new BinaryExpression(Token.LOGICAL_OR, expression, parseOrExpression());
		}

		return expression;
	}

	private Expression parseAndExpression() {
		Expression expression = parseBitwiseOrExpression();

		if (accept(Token.LOGICAL_AND)) {
			return new BinaryExpression(Token.LOGICAL_AND, expression, parseAndExpression());
		}

		return expression;
	}

	private Expression parseBitwiseOrExpression() {
		Expression expression = parseBitwiseXorExpression();

		if (accept(Token.BITWISE_OR)) {
			return new BinaryExpression(Token.BITWISE_OR, expression, parseBitwiseOrExpression());
		}

		return expression;
	}

	private Expression parseBitwiseXorExpression() {
		Expression expression = parseBitwiseAndExpression();

		if (accept(Token.BITWISE_XOR)) {
			return new BinaryExpression(Token.BITWISE_XOR, expression, parseBitwiseXorExpression());
		}

		return expression;
	}

	private Expression parseBitwiseAndExpression() {
		Expression expression = parseEqualityExpression();

		if (accept(Token.BITWISE_AND)) {
			return new BinaryExpression(Token.BITWISE_AND, expression, parseBitwiseAndExpression());
		}

		return expression;
	}

	private Expression parseEqualityExpression() {
		Expression expression = parseRelationalExpression();

		if (accept(Token.LOGICAL_EQUAL, Token.LOGICAL_NOT_EQUAL)) {
			return new BinaryExpression(current.getToken(), expression, parseEqualityExpression());
		}

		return expression;
	}

	private Expression parseRelationalExpression() {
		Expression expression = parseBitwiseShiftExpression();

		if (accept(RELATIONAL_OPERATORS)) {
			return new BinaryExpression(current.getToken(), expression, parseRelationalExpression());
		}

		return expression;
	}

	private Expression parseBitwiseShiftExpression() {
		Expression expression = parseAdditiveExpression();

		if (accept(Token.BITWISE_LEFT_SHIFT, Token.ASSIGN_BITWISE_RIGHT_SHIFT, Token.ASSIGN_BITWISE_UNSIGNED_RIGHT_SHIFT)) {
			return new BinaryExpression(current.getToken(), expression, parseBitwiseShiftExpression());
		}

		return expression;
	}

	private Expression parseAdditiveExpression() {
		Expression expression = parseMultiplicativeExpression();

		if (accept(Token.PLUS, Token.MINUS)) {
			return new BinaryExpression(current.getToken(), expression, parseAdditiveExpression());
		}

		return expression;
	}

	private Expression parseMultiplicativeExpression() {
		Expression expression = parseUnaryExpression();

		if (accept(Token.ASTERISK, Token.SLASH, Token.PERCENT)) {
			return new BinaryExpression(current.getToken(), expression, parseMultiplicativeExpression());
		}

		return expression;
	}

	private Expression parseUnaryExpression() {
		if (accept(Token.INCREMENT, Token.DECREMENT)) {
			Token token = current.getToken();
			Expression expression = parseUnaryExpression();

			if (!isLeftHandSide(expression)) {
				throw ParserException.illegalLeftHandSide(current.getPosition());
			}

			return new UnaryExpression(token, expression, UnaryExpression.Kind.PREFIX);
		}

		if (accept(Token.MINUS, Token.BITWISE_NOT, Token.LOGICAL_NOT, Token.KEYWORD_DELETE)) {
			Token token = current.getToken();
			Expression expression = parseUnaryExpression();

			return new UnaryExpression(token, expression, UnaryExpression.Kind.PREFIX);
		}

		return parsePostfixExpression();
	}

	private Expression parsePostfixExpression() {
		Expression expression = parseLeftHandSideExpressionAllowCall();

		while (true) {
			if (match(Token.INCREMENT, Token.DECREMENT) && next.getToken() != Token.SEMI_COLON) {
				if (!isLeftHandSide(expression)) {
					throw ParserException.illegalLeftHandSide(current.getPosition());
				}

				accept(Token.INCREMENT, Token.DECREMENT);

				expression = new UnaryExpression(current.getToken(), expression, UnaryExpression.Kind.POSTFIX);
			} else {
				break;
			}
		}

		return expression;
	}

	private Expression parseLeftHandSideExpression() {
		Expression expression = parsePrimaryExpression();

		while (match(Token.DOT, Token.OPEN_BRACKET)) {
			if (match(Token.OPEN_BRACKET)) {
				Expression member = parseComputedMemberExpression();
				expression = new MemberExpression(MemberExpression.Kind.LIST, expression, member);
			} else {
				Expression property = parseNonComputedMemberExpression();
				expression = new MemberExpression(MemberExpression.Kind.PROPERTY, expression, property);
			}
		}

		return expression;
	}

	private Expression parseLeftHandSideExpressionAllowCall() {
		Expression expression = parsePrimaryExpression();

		while (match(Token.DOT, Token.OPEN_BRACKET, Token.OPEN_PARENTHESIS)) {
			if (match(Token.OPEN_PARENTHESIS)) {
				List<Expression> arguments = parseArguments();

				expression = new CallExpression(expression, arguments);
			} else if (match(Token.OPEN_BRACKET)) {
				Expression member = parseComputedMemberExpression();
				expression = new MemberExpression(MemberExpression.Kind.LIST, expression, member);
			} else {
				Expression property = parseNonComputedMemberExpression();
				expression = new MemberExpression(MemberExpression.Kind.PROPERTY, expression, property);
			}
		}

		return expression;
	}

	private Expression parseComputedMemberExpression() {
		check(Token.OPEN_BRACKET);
		Expression expression = parseExpression();
		check(Token.CLOSE_BRACKET);

		return expression;
	}

	private Expression parseNonComputedMemberExpression() {
		check(Token.DOT);

		return parseNonComputedPropertyExpression();
	}

	private Expression parseNonComputedPropertyExpression() {
		check(Token.IDENTIFIER);

		return new Identifier(current.getLexeme());
	}

	private Expression parsePrimaryExpression() {
		if (accept(Token.OPEN_PARENTHESIS)) {
			Expression expression = parseExpression();
			check(Token.CLOSE_PARENTHESIS);

			return expression;
		}

		if (accept(Token.IDENTIFIER)) {
			return new Identifier(current.getLexeme());
		}

		if (accept(Token.KEYWORD_THIS)) {
			return new ReferenceExpression(ReferenceExpression.Kind.THIS);
		}

		if (accept(Token.KEYWORD_BASE)) {
			return new ReferenceExpression(ReferenceExpression.Kind.BASE);
		}

		if (accept(Token.LITERAL_INTEGER)) {
			return new Literal(Literal.Kind.INTEGER, Long.parseLong(current.getLexeme()));
		}

		if (accept(Token.LITERAL_FLOAT)) {
			return new Literal(Literal.Kind.FLOAT, Double.parseDouble(current.getLexeme()));
		}

		if (accept(Token.LITERAL_STRING)) {
			return new Literal(Literal.Kind.STRING, current.getLexeme());
		}

		if (accept(Token.LITERAL_BOOLEAN)) {
			return new Literal(Literal.Kind.BOOLEAN, current.getLexeme().equals("true"));
		}

		if (accept(Token.KEYWORD_NIL)) {
			return new Literal(Literal.Kind.NIL, null);
		}

		if (match(Token.DOLLAR)) {
			return parseFunctionExpression();
		}

		if (match(Token.HASH)) {
			return parseClassExpression();
		}

		if (match(Token.OPEN_BRACE)) {
			return parseObjectInitializer();
		}

		throw ParserException.internal();
	}

	private Expression parseFunctionExpression() {
		check(Token.DOLLAR, Token.OPEN_PARENTHESIS);
		List<Identifier> parameters = parseParameters();
		check(Token.CLOSE_PARENTHESIS);
		Statement body = parseStatement();

		return new FunctionExpression(parameters, body);
	}

	private Expression parseClassExpression() {
		return null;
	}

	private Expression parseObjectInitializer() {
		return null;
	}

	private List<Expression> parseArguments() {
		List<Expression> arguments = new ArrayList<Expression>();

		while (true) {
			if (match(Token.CLOSE_PARENTHESIS)) {
				break;
			}

			arguments.add(parseAssignmentExpression());

			if (!accept(Token.COMMA)) {
				break;
			}
		}

		return arguments;
	}

	private List<Identifier> parseParameters() {
		List<Identifier> parameters = new ArrayList<Identifier>();

		while (accept(Token.IDENTIFIER)) {
			parameters.add(new Identifier(current.getLexeme()));

			if (!accept(Token.COMMA)) {
				break;
			}
		}

		return parameters;
	}

	private boolean match(Token token) {
		return next.getToken() == token;
	}

	private boolean match(Token... tokens) {
		for (int i = 0; i < tokens.length; i++) {
			if (match(tokens[i])) {
				return true;
			}
		}

		return false;
	}

	private void check(Token token) {
		if (match(token)) {
			current = next;
			next = lexer.next();
		} else {
			throw ParserException.unexpected(current.getPosition(), token, next.getToken());
		}
	}

	private void check(Token... tokens) {
		for (int i = 0; i < tokens.length; i++) {
			check(tokens[i]);
		}
	}

	private boolean accept(Token token) {
		if (match(token)) {
			current = next;
			next = lexer.next();

			return true;
		}

		return false;
	}

	private boolean accept(Token... tokens) {
		for (int i = 0; i < tokens.length; i++) {
			if (accept(tokens[i])) {
				return true;
			}
		}

		return false;
	}

	private boolean isLeftHandSide(Expression expression) {
		return expression instanceof Identifier || expression instanceof MemberExpression;
	}

}
